syntax = "proto3";

package library;

service LibraryService {

  // -------- AUTH --------
  rpc UserLogin (LoginRequest) returns (LoginResponse);

  // -------- BOOK --------
  rpc CreateBook (Book) returns (StatusResponse);
  rpc SearchBooks (SearchRequest) returns (stream Book);

  // -------- USERS --------
  rpc GetAllUsers (SearchRequest) returns (stream UserDetail);
  rpc GetUserDetail (UserIdRequest) returns (UserDetail);
  rpc DeleteUser (UserIdRequest) returns (StatusResponse);

  // -------- CLIENTS --------
  rpc CreateClient (Client) returns (StatusResponse);
  rpc GetAllClients (SearchRequest) returns (stream Client);
  rpc GetClient (ClientIdRequest) returns (Client);
  rpc UpdateClient (Client) returns (StatusResponse);
  rpc DeleteClient (ClientIdRequest) returns (StatusResponse);
}

// -------- MESSAGES --------

message LoginRequest {
  string username = 1;
  string password = 2;
}

message LoginResponse {
  bool success = 1;
  string user_id = 2;
  string message = 3;
}

message SearchRequest {
  string query = 1;
}

message UserIdRequest {
  string user_id = 1;
}

message ClientIdRequest {
  int32 client_id = 1;
}

message StatusResponse {
  bool success = 1;
  string message = 2;
  int32 entity_id = 3;
}

message UserDetail {
  string user_id = 1;
  string username = 2;
  string email = 3;
  bool is_staff = 4;
  bool is_active = 5;
  bool is_superuser = 6;
  string date_joined = 7;
}

message Book {
  int32 id = 1;
  string title = 2;
  string author = 3;
  string isbn = 4;
  int32 total_copies = 5;
  int32 available_copies = 6;
  string image_url = 7;
}

message Client {
  int32 id = 1;
  string nom = 2;
  string email = 3;
  string telephone = 4;
  string adresse = 5;
  string date_inscription = 6;
}

// Pour g√©rer les emprunts
message BorrowRequest {
    string client_id = 1;
    string book_id = 2;
    string borrow_date = 3;
    string return_date = 4;
}

message BorrowResponse {
    bool success = 1;
    string message = 2;
}

message BorrowListRequest {
    string client_id = 1;
}

message BorrowListResponse {
    string id = 1;
    string book_title = 2;
    string borrow_date = 3;
    string return_date = 4;
}
